relax litellm provider constraint

arjun-krishna1 opened the pull request:
Fixing: https://github.com/leondz/garak/issues/755
- Remove provider constraint on litellm generator
- Fixes breaking garak test `test_litellm#test_litellm_openai`
- Exceptions on model non-existence raised by litellm bubbled up

arjun-krishna1 commented:
- `test_litellm#test_litellm_openai` passes on this branch but fails on main
 ```bash
 ~/garak$ git branch
* bugfix/litellm_provider_validation
  main
~/garak$ python -m pytest tests/generators/test_litellm.py::test_litellm_openai -s
======================================================= test session starts =======================================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/arjun/garak
configfile: pyproject.toml
plugins: requests-mock-1.12.1, anyio-4.4.0, respx-0.21.1
collected 1 item                                                                                                                  

tests/generators/test_litellm.py 🦜 loading generator: LiteLLM: gpt-3.5-turbo
test passed!
.

======================================================== 1 passed in 3.98s ========================================================
```
- Fails on main:
```bash
~/garak$ python -m pytest tests/generators/test_litellm.py::test_litellm_openai -s
======================================================= test session starts =======================================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/arjun/garak
configfile: pyproject.toml
plugins: requests-mock-1.12.1, anyio-4.4.0, respx-0.21.1
collected 1 item                                                                                                                  

tests/generators/test_litellm.py 🦜 loading generator: LiteLLM: gpt-3.5-turbo
F

============================================================ FAILURES =============================================================
_______________________________________________________ test_litellm_openai _______________________________________________________

    @pytest.mark.skipif(
        getenv("OPENAI_API_KEY", None) is None,
        reason="OpenAI API key is not set in OPENAI_API_KEY",
    )
    def test_litellm_openai():
        model_name = "gpt-3.5-turbo"
>       generator = LiteLLMGenerator(name=model_name)

tests/generators/test_litellm.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <garak.generators.litellm.LiteLLMGenerator object at 0x7d1108f95d30>, name = 'gpt-3.5-turbo', generations = 10
config_root = <module 'garak._config' from '/home/arjun/garak/garak/_config.py'>

    def __init__(self, name: str = "", generations: int = 10, config_root=_config):
        self.name = name
        self.api_base = None
        self.api_key = None
        self.provider = None
        self.key_env_var = self.ENV_VAR
        self.generations = generations
        self._load_config(config_root)
        self.fullname = f"LiteLLM {self.name}"
        self.supports_multiple_generations = not any(
            self.name.startswith(provider)
            for provider in unsupported_multiple_gen_providers
        )
    
        super().__init__(
            self.name, generations=self.generations, config_root=config_root
        )
    
        if self.provider is None:
>           raise ValueError(
                "litellm generator needs to have a provider value configured - see docs"
E               ValueError: litellm generator needs to have a provider value configured - see docs

garak/generators/litellm.py:129: ValueError
===================================================== short test summary info =====================================================
FAILED tests/generators/test_litellm.py::test_litellm_openai - ValueError: litellm generator needs to have a provider value configured - see docs
======================================================== 1 failed in 1.06s ========================================================
```

arjun-krishna1 commented:
Exception on non-existence raised by litellm:
```py
>>> from garak.generators.litellm import LiteLLMGenerator
>>> non_existent_model = "non-existent-model"
>>> generator = LiteLLMGenerator(name=non_existent_model)
🦜 loading generator: LiteLLM: non-existent-model
>>> generator.generate("This should raise an exception!")

Provider List: https://docs.litellm.ai/docs/providers

INFO:backoff:Backing off _call_model(...) for 0.0s (litellm.exceptions.BadRequestError: litellm.BadRequestError: LLM Provider NOT provided. Pass in the LLM provider you are trying to call. You passed model=non-existent-model
 Pass model as E.g. For 'Huggingface' inference endpoints pass in `completion(model='huggingface/starcoder',..)` Learn more: https://docs.litellm.ai/docs/providers)
 ```

arjun-krishna1 commented:
I have read the DCO Document and I hereby sign the DCO

leondz commented:
Thank you, will take a look

jmartin-tech commented:
@arjun-krishna1, please follow the fine print at the end of the bot's DCO comment to trigger action again.

jmartin-tech reviewed (commented):
This is definitely an improvement, and it exposes another layer that needs to be accounted for.

Based on this change if no provider is passed the `litellm.completion()` preformed in `_call_model()` will enter an infinite backoff when it raises exceptions for a missing `api_key` value depending on what provider it `autodetects` from the model `name` provided.
 
A try block around `litellm.completion()` is need that will capture `litellm.exceptions.AuthenticationError` and possibly raise `garak.exception.BadGeneratorException` from the original error to cause the run to exit. Also the `@backoff.on_exception` would need to more explicitly only backoff on `litellm.exceptions.APIError` instead of any raised `Exception`.

arjun-krishna1 commented:
Thanks for the review @jmartin-tech
Updated PR with try block and only backing off on litellm APIError
Have also added a test that checks that a Bad Generator exception is raised if litellm is given a bad model name

jmartin-tech commented on tests/generators/test_litellm.py:
```diff
@@ -43,3 +44,11 @@ def test_litellm_openrouter():
     print("test passed!")
+
+
+def test_litellm_model_non_existence():
+    model_name = "non-existent-model"
+    generator = LiteLLMGenerator(name=model_name)
+    with pytest.raises(BadGeneratorException):
+        output = generator.generate("This should raise an exception")
+    assert "Exceptions on model non-existence raised by litellm should be bubbled up"
```
Since there is no actual `litellm` instance reached by this test, I would expect there needs to be more configuration values passed to call the completion.

This test would need to be passed `api_base` and also a model name that can trigger `openai` detection:
```suggestion
def test_litellm_model_detection():
    custom_config = {
        "generators": {
            "litellm": {
                "api_base": "https://garak.example.com/v1",
            }
        }
    }
    model_name = "non-existent-model"
    generator = LiteLLMGenerator(name=model_name, config_root=custom_config)
    with pytest.raises(BadGeneratorException):
        generator.generate("This should raise an exception")
    generator = LiteLLMGenerator(name="openai/invalid-model", config_root=custom_config)
    with pytest.raises(BadGeneratorException):
        generator.generate("This should raise an exception")
```

The assert at end of the test is redundant although capturing the exception and asserting some unique expectation on it would make sense.

arjun-krishna1 replied:
This new test fails if we don't catch BadRequestError
litellm raises BadRequestError if we give a non existent model without a provider

```
E               litellm.exceptions.BadRequestError: litellm.BadRequestError: LLM Provider NOT provided. Pass in the LLM provider you are trying to call. You passed model=non-existent-model
E                Pass model as E.g. For 'Huggingface' inference endpoints pass in `completion(model='huggingface/starcoder',..)` Learn more: https://docs.litellm.ai/docs/providers

.venv/lib/python3.12/site-packages/litellm/utils.py:4827: BadRequestError
============================================================================================ short test summary info =============================================================================================
FAILED tests/generators/test_litellm.py::test_litellm_model_detection - litellm.exceptions.BadRequestError: litellm.BadRequestError: LLM Provider NOT provided. Pass in the LLM provider you are trying to call. You passed model=non-existent-model
```

jmartin-tech replied:
In my local tests an model without a provider detected would raise, `APIConnectionError`. The upstream library has changed behavior: 

```
litellm.exceptions.APIConnectionError: LLM Provider NOT provided. Pass in the LLM provider you are trying to call. You passed model=non-existent-model
 Pass model as E.g. For 'Huggingface' inference endpoints pass in `completion(model='huggingface/starcoder',..)` Learn more: https://docs.litellm.ai/docs/providers
```

As of `1.41.21` the raised error type changed to BadRequestError, please update `requirements.txt` and `pyproject.toml` to required `litellm>=1.41.21` and adjust the `try` block to catch (BadRequestError, AuthenticationError) to match the newer library package expectations.

arjun-krishna1 replied:
Thanks for this investigation @jmartin-tech
Updated the litellm version and adjusted the try block

jmartin-tech reviewed (changes requested):
To complete this constraint removal, the code needs to fully support passing model type detection on to `litellm`. This means removing the class level `ENV_VAR` and raising for specific errors that are thrown when `litellm` cannot determine the target API client to utilize.

arjun-krishna1 commented on garak/generators/litellm.py:
```diff
@@ -125,11 +126,7 @@ def __init__(self, name: str = "", generations: int = 10, config_root=_config):
-                "litellm generator needs to have a provider value configured - see docs"
-            )
-        elif (
+        if (
```
Updated

arjun-krishna1 commented on garak/generators/litellm.py:
```diff
@@ -157,20 +154,23 @@ def _call_model(
+                custom_llm_provider=self.provider,
+                api_key=self.api_key,
+            )
+        except litellm.exceptions.BadRequestError as e:
```
@jmartin-tech when giving a non existent model without a provider, litellm raises a BadRequestError with this message:
```
FAILED tests/generators/test_litellm.py::test_litellm_model_non_existence -
litellm.exceptions.BadRequestError: litellm.BadRequestError: LLM Provider NOT provided. Pass in the LLM provider you are trying to call. You passed model=non-existent-model
```

jmartin-tech replied:
My comment should have been superseded by a [later one](https://github.com/leondz/garak/pull/820#discussion_r1720219972) catching (AuthenticationError, APIConnectionError) the not detected error is being `APIConnectionError` handled in my local env when, it looks like this changed in the `litellm` packages. As of `1.41.21` the raised error type changed to `BadRequestError`.

jmartin-tech commented on garak/generators/litellm.py:
```diff
@@ -96,7 +92,6 @@ class LiteLLMGenerator(Generator):
         "generations",
         "context_len",
         "max_tokens",
-        "api_key",
```
This should still be a `supported_param`, when `_supported_params` is defined any parameter not listed is filtered from configuration files.

arjun-krishna1 replied:
Added api key back as a supported param @jmartin-tech

arjun-krishna1 commented:
> To complete this constraint removal, the code needs to fully support passing model type detection on to `litellm`. This means removing the class level `ENV_VAR` and raising for specific errors that are thrown when `litellm` cannot determine the target API client to utilize.

Hi @jmartin-tech , I think I've resolved all your comments so far
Please let me know if you have any other feedback or if this is good to go!

jmartin-tech commented on garak/generators/litellm.py:
```diff
@@ -157,20 +136,26 @@ def _call_model(
+            litellm.exceptions.BadRequestError,
+        ) as e:
+
+            raise BadGeneratorException() from e
```
```suggestion
            raise BadGeneratorException("Unrecoverable error during litellm completion see log for details") from e
```

jmartin-tech reviewed (approved):
Testing shows one more minor UX improvement need.

```
$ python -m garak -m litellm -n openai/meta/llama3-8b-instruct -g 1 -p continuation --generator_option_file litellm.json 2> /dev/null
garak LLM vulnerability scanner v0.9.0.14.post1 ( https://github.com/leondz/garak ) at 2024-08-22T10:36:48.688446
📜 logging to /home/jemartin/.local/share/garak/garak.log
🦜 loading generator: LiteLLM: openai/meta/llama3-8b-instruct
📜 reporting to /home/jemartin/.local/share/garak/garak_runs/garak.703878b2-c57d-49f2-8af0-bff238965aab.report.jsonl
🕵️  queue of probes: continuation.ContinueSlursReclaimedSlursMini

Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

```

The raised `BadGeneratorException` can report the message used for instantiation to guide user on how to determine `why` the failure occurred. Committing a suggested error message and landing this shortly. 

```
$ python -m garak -m litellm -n openai/meta/llama3-8b-instruct -g 1 -p continuation --generator_option_file litellm.json 2> /dev/null
garak LLM vulnerability scanner v0.9.0.14.post1 ( https://github.com/leondz/garak ) at 2024-08-22T10:37:55.025515
📜 logging to /home/jemartin/.local/share/garak/garak.log
🦜 loading generator: LiteLLM: openai/meta/llama3-8b-instruct
📜 reporting to /home/jemartin/.local/share/garak/garak_runs/garak.3838265e-c493-4bc7-a5de-0a047d016900.report.jsonl
🕵️  queue of probes: continuation.ContinueSlursReclaimedSlursMini

Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

Unrecoverable error during litellm completion see log for details
```

leondz commented:
Thank you @arjun-krishna1 !!
